src/v/serde/rw/enum.h

```diff
@@ -24,6 +24,7 @@ namespace serde {
 template<typename T>
 requires(serde_is_enum_v<std::decay_t<T>>)
 void tag_invoke(tag_t<write_tag>, iobuf& out, T t) {
+    static_assert(std::is_scoped_enum_v<std::decay_t<T>>);
```

Comment:
don't let your friends use non-scoped enums

Reply:
`template< class T >struct is_scoped_enum;    (since C++23)` will not work with CMake :-(

Reply:
it works with Bazel, and when I tried to switch cmake to c++23 it broke the upgrade tests because this was an incompatible change to the serialization format for this enum

Reply:
oh I see, we need to guard the static assert with c++ version checks

Reply:
Maybe write in a more verbose way that would work with both c++20 and c++23 until we fully switch? [Something like this](https://en.cppreference.com/w/cpp/types/is_scoped_enum#Possible_implementation).

Reply:
i don't think we need to do anything here, right? we can run CI on the vtools side with c++23 turned on?

Reply:
Yeah but that's going to require the annoying commit vtools PR and this PR "together" song and dance. I'm fine with doing that (glad that we don't have to do this anymore with Bazel).

Reply:
LOL i literally used the phrasing "song and dance" in the PR i just pushed which made the same observation

Reply:
> Maybe write in a more verbose way that would work with both c++20 and c++23 until we fully switch? [Something like this](https://en.cppreference.com/w/cpp/types/is_scoped_enum#Possible_implementation).

Good idea. I think we can compile it out of the tree for now, though, because in principle the fix only needs to last a day or two until vtools switches over to c++23.